helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -1278,29 +1276,28 @@ private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry>
             + listener.getDataListener() + " prefetch data: " + listener.isPrefetchData()) {
           @Override
           public void run() throws Exception {
-            // Reinstall watch before listener callbacks to check the znode status
+            if (!pathExists) {
+              listener.getDataListener().handleDataDeleted(path);
+              return;
+            }
             if (!pathStatRecord.pathChecked()) {
               pathStatRecord.recordPathStat(getStat(path, true), notificationTime);
```

Comment:
You avoided the notificationTime report on the deleting events with this change. Please keep the original order.

Reply:
This will invalidate our propagation latency metric, BTW.

Reply:
Thanks for catching it, fixed it.